enhancements/substitutes-for.md

```diff
@@ -0,0 +1,379 @@
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+
+#### Freshmaker rebuilds a bundle that is the head of the channel, the bundle
+does not exist in any other channels.
+
+In this case, freshmaker simply takes the CSV it is rebuilding and adds a
+"skips" reference naming the bundle it is rebuilding.  A substitutes-for
+reference must also be added to the CSV.
+
+#### Bundle B replaces bundle A, bundle A is rebuilt as A' to fix a CVE.
+
+Freshmaker will rebuild A as A', adding a "skips: A" reference.  In addition
+it will add a "substitutes-for: A" reference.
+
+When ListBundles serves the CSV for B, B will include the following references:
+1) replaces A (from the original CSV)
+2) skips A' (because of replaces A plus the A->A' substitution rule).  Ideally
+this would be a "replaces" reference, but we cannot have multiple replaces
+references.
+
+The result of this is that users on A can update to A' or B(depending which
+channel they are on).  Users on A' can update to B.
+
+**Note: This is not ideal since it means a user can be upgraded from something older
+than A/A', directly to B, without going through A/A', which may mean they
+skip a necessary migration.  Need to discuss this scenario further**
+
+#### Bundle B skips bundle A, bundle A is rebuilt as A' to fix a CVE.
+
+Freshmaker will rebuild A as A', adding a "skips: A" reference.  In addition
+it will add a "substitutes-for: A" reference.
+
+When ListBundles serves the CSV for B, B will include the following references:
+1) skips A (from the original CSV)
+2) skips A' (because of skips A plus the A->A' substitution rule).
+
+The result of this is that users on A can update to A' or B(depending which
+channel they are on).  Users on A' can update to B.
+
+
+#### Bundle A is rebuilt as A', bundle A' is rebuilt as A''
+
+Substitution rules are transitive, so ListBundles must mutate any references
+(replaces/skips) to A such that they also include skips of A' and A''.
+
+Again ideally an existing `replaces A` would be updated to `replaces: A,A',A''`
+but since multiple replaces are not valid, we must settle for `replaces: A`
+plus `skips: A',A''`.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+The primary caveat with this design is how to handle the need for updating
+a single "replaces" reference to now reference multiple bundles.  For
+now this design proposes converting the additional replaces references
+into "skips", but there is an argument for instead moving the existing
+"replaces" reference to be a skips, and making the "replaces" reference
+point to the "newest" substitution.  That would at least ensure that
+if/when we must install a bundle as part of the path to upgrade to the
+latest, we install the most recently patched version of that bundle.
+
+However it adds additional complexity to determine which bundle is the
```

Comment:
Why is it difficult to determine the latest rebuild?

Reply:
"replaces" is a reference to a csv name, not a version, so there's no guaranteed semantic ordering.

Suppose B replaces A.  If i rebuilt A as A.foo and then i rebuild A again as A.bar.  Which one goes into the replaces field?

In theory for freshmaker's use case we might be able to make assumptions about versions and ordering that allow us to resolve this, but in the generic case I don't think we have the information needed to be certain.

Reply:
Because the graph is always a linear path of replaces (adding skips is basically orthogonal to that), it does stand to reason that as long as `substitutes-for` has to also be unique for the graph, you could fairly trivially reconstruct this metadata every time -- you would just need to also keep track of what the original replaces and substitutes-for data is for each of the bundles.

That being said, if we're saying that `substitutes-for` does *not* have to be unique, then we have a temporal problem and this becomes non-deterministic.

Reply:
I think `substitutes-for` should be unique to the graph. If a newer substitution is added, then it should be against the bundle image that did the original substitution.

Reply:
Yes i think we can assert that only one thing can substitute for another, and if you got it wrong, you introduce a third thing that substitutes for the second thing.

I will make that restriction explicit.

So think that means we can assert that we know the ultimate resolution of any chain of substitutions, and therefore can modify the "Replaces" field.  so i'll update it to reflect that as well.

Reply:
i've made this update, so the behavior is now to mutate the replaces field, and then add any additional skips that are needed to cover the substituted versions.